### PR TITLE
improve the logic for error collection

### PR DESCRIPTION
**Pull request by rakesh-garimella**
**Description**

Changes proposed in this pull request:

-Aggregate errors from both CR update and fluent-bit restart and then requeue request if required.

**Related issue(s)**

https://github.com/kyma-project/kyma/issues/15199

### CODE REVIEW

File: components/telemetry-operator/controllers/telemetry/logparser_controller.go [hunk 1]
```diff
@@ -72,18 +72,26 @@ func (r *LogParserReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
 		return ctrl.Result{Requeue: shouldRetryOn(err)}, nil
 	}
 
+	var errMap = make(map[string]error)
+
 	if result.LogParserChanged {
 		if err = r.Update(ctx, &logParser); err != nil {
-			log.Error(err, "Failed updating log parser")
-			return ctrl.Result{Requeue: shouldRetryOn(err)}, err
+			errMap["Failed updating log parser"] = err
```
Comment (chrkl) [inline components/telemetry-operator/controllers/telemetry/logparser_controller.go:79]: You never return this error if the condition in line 83 does not match

Comment (rakesh-garimella) [reply components/telemetry-operator/controllers/telemetry/logparser_controller.go:84]: good catch fixed that

File: components/telemetry-operator/controllers/telemetry/logparser_controller.go [hunk 2]
```diff
@@ -122,6 +116,28 @@ func (r *LogParserReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
 	return ctrl.Result{}, nil
 }
 
+func (r *LogParserReconciler) updateCR(ctx context.Context, updateCR bool, logParser telemetryv1alpha1.LogParser) error {
+	if !updateCR {
+		return nil
+	}
+	if err := r.Update(ctx, &logParser); err != nil {
```
Comment (chrkl) [inline components/telemetry-operator/controllers/telemetry/logparser_controller.go:123]: Just make a `return r.Update ...` (here and below)

File: components/telemetry-operator/controllers/telemetry/logpipeline_controller.go [hunk 3]
```diff
@@ -186,6 +179,38 @@ func shouldRetryOn(err error) bool {
 		!errors.IsNotFound(err)
 }
 
+func getError(crUpdateError, fbRestartError error) error {
```
Comment (chrkl) [inline components/telemetry-operator/controllers/telemetry/logpipeline_controller.go:182]: Let's make this the following: `func getFirstError(err ...error) error`
The name describes better what is actually happening and you could just return the first list element (or nil). Maybe let's move this to a new `error_helpers` file together with `shouldRetryOn` to do not use code from a different controller.

File: components/telemetry-operator/controllers/telemetry/logparser_controller.go [hunk 4]
```diff
@@ -122,6 +116,28 @@ func (r *LogParserReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
 	return ctrl.Result{}, nil
 }
 
+func (r *LogParserReconciler) updateCR(ctx context.Context, updateCR bool, logParser telemetryv1alpha1.LogParser) error {
```
Comment (chrkl) [inline components/telemetry-operator/controllers/telemetry/logparser_controller.go:119]: Can we call it `updateLogParser` (and `updateLogPipeline` in the other controller)? The meaning of the shortcut CR is not very obvious.
Maybe we should also indicate that the behaviour is conditional: `updateLogParserIfChanged`

**Review by chrkl — CHANGES_REQUESTED** [review]

**Comment by chrkl** [comment]
We will not continue with the proposed changes since it does not cover all corner cases for the Fluent Bit restart properly.

**Pull request closed without merging**